presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/BaseCostBasedPlanTest.java

```diff
@@ -0,0 +1,248 @@
+import static org.testng.Assert.assertEquals;
+
+public abstract class BaseCostBasedPlanTest
+        extends BasePlanTest
+{
+    public BaseCostBasedPlanTest(LocalQueryRunnerSupplier supplier)
+    {
+        super(supplier);
+    }
+
+    protected abstract Stream<Query> getQueries();
+
+    protected abstract Path getExpectedJoinOrderingFile(String queryId);
+
+    @DataProvider
+    public Object[][] getQueriesDataProvider()
+    {
+        return getQueries()
+                .collect(toDataProvider());
+    }
+
+    @Test(dataProvider = "getQueriesDataProvider")
+    public void test(Query query)
+    {
+        assertEquals(generateQueryPlan(query.getQuery()), read(resolve(getExpectedJoinOrderingFile(query.getId()))));
+    }
+
+    private Path resolve(Path path)
+    {
+        // Test is run with module as current working directory. generate() is run with project's top-level directory as current working directory.
+        if (exists(path)) {
+            return path;
+        }
+        Path fromParent = Paths.get("..").resolve(path);
+        checkState(exists(fromParent), "Unable to resolve: " + path);
+        return fromParent;
+    }
+
+    public void generate()
+            throws Exception
+    {
+        initPlanTest();
+        try {
+            getQueries()
+                    .parallel()
+                    .forEach(query -> {
+                        try {
+                            Files.write(generateQueryPlan(query.getQuery()).getBytes(UTF_8), getExpectedJoinOrderingFile(query.id).toFile());
+                            System.out.println(query.getId());
+                        }
+                        catch (IOException e) {
+                            throw new UncheckedIOException(e);
+                        }
+                    });
+        }
+        finally {
+            destroyPlanTest();
+        }
+    }
+
+    private static String read(Path file)
+    {
+        try {
+            return Files.asCharSource(file.toFile(), UTF_8).read();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public static final class Query
+    {
+        private final String id;
+        private final String query;
+
+        public Query(String id, String query)
+        {
+            this.id = requireNonNull(id, "id is null");
+            this.query = requireNonNull(query, "query is null");
+        }
+
+        public String getId()
+        {
+            return id;
+        }
+
+        public String getQuery()
+        {
+            return query;
+        }
+
+        @Override
+        public String toString()
+        {
+            // Keep toString() short, as this is used in @DataProvider
+            return id;
+        }
+    }
+
+    private String generateQueryPlan(String query)
+    {
+        String sql = query.replaceFirst(";", "")
+                .replace("${database}.${schema}.", "")
```

Comment:
What about a regex? `(\"?\$\{database\}\"?\\.)?(\"?\$\{schema\}\"?\\.)?\.\"?\$\{prefix\}\"?`?

Reply:
I do not feel that this regex is better. It much more complicated and does not solve all the issues here. Please notice that we have to rules, one removes all the occurences while the other one replace occurences with`"`